src/HotChocolate/Core/src/Execution/Utilities/ObjectBuffer.cs

```diff
@@ -33,25 +29,84 @@ public T Pop()
 
         public bool TryPop([NotNullWhen(true)] out T? obj)
         {
-            if (_index < _capacity)
+            var nextIndex = _index++;
+            if (nextIndex <= _capacity)
             {
-                obj = _buffer[_index++];
+                obj = _buffer[_index] ?? new T();
+                _buffer[_index] = null;
                 return true;
             }
 
             obj = null;
             return false;
         }
 
-        public void Reset()
+        public T PopSafe()
+        {
+            if (TryPopSafe(out T? obj))
+            {
+                return obj;
+            }
+            throw new InvalidOperationException("Buffer is used up.");
+        }
+
+        public bool TryPopSafe([NotNullWhen(true)] out T? obj)
+        {
+            var nextIndex = Interlocked.Increment(ref _index);
+            if (nextIndex <= _capacity)
+            {
+                obj = _buffer[_index] ?? new T();
```

Comment:
you have to use the nextIndex here since the other threads might have increased _index already.

Reply:
we could even change this to start with -1 so you can just inrement... in this case we would need to change the standard pop to `++_index`

Reply:
But in this case we also would need to change the decrement .... so better to leave it like it is and do a decrement on nextIndex

Reply:
haha good catch

Reply:
this should obviously be [nextIndex]